File: src/main/java/io/funky/fangs/keep_it_personal/command/KeepItPersonalCommand.java
package io.funky.fangs.keep_it_personal.command;

import com.mojang.brigadier.Command;
import com.mojang.brigadier.context.CommandContext;
import io.funky.fangs.keep_it_personal.domain.DeathPreferenceContainer;
import net.minecraft.server.command.ServerCommandSource;

import static io.funky.fangs.keep_it_personal.KeepItPersonalModInitializer.MOD_ID;
import static net.minecraft.text.Text.translatable;

public abstract class KeepItPersonalCommand implements Command<ServerCommandSource> {
    public static final int SINGLE_ERROR = -1;
    public static final String NAME = "kip";
    public static final String PERMISSION = MOD_ID + '.' + NAME;
    private static final String INVALID_COMMAND_SENDER = "keep-it-personal.message.invalid-command-sender";

    @Override
    public int run(CommandContext<ServerCommandSource> context) {
        final var source = context.getSource();
        final var player = source.getPlayer();

        if (player instanceof DeathPreferenceContainer container) {
            final var result = run(context, container);
            return result == null ? SINGLE_ERROR : result;
        }
        else {
            source.sendError(translatable(INVALID_COMMAND_SENDER));
            return SINGLE_ERROR;
        }
    }

    public abstract Integer run(CommandContext<ServerCommandSource> context, DeathPreferenceContainer container);
}

File: src/main/kotlin/io/funky/fangs/keep_it_personal/command/KeepItPersonalCommand.kt
package io.funky.fangs.keep_it_personal.command

import com.mojang.brigadier.Command
import com.mojang.brigadier.context.CommandContext
import net.minecraft.server.command.ServerCommandSource
import net.minecraft.server.network.ServerPlayerEntity
import net.minecraft.text.Text

const val KEEP_COMMAND = "keeping"
const val ARGUMENT_NAME = "shouldKeep"
const val SINGLE_ERROR = -1

abstract class KeepItPersonalCommand: Command<ServerCommandSource> {
    override fun run(context: CommandContext<ServerCommandSource>): Int {
        val source = context.source
        val server = source.server
        val player = source.player

        if (player != null) {
            val keepCommandState = KeepCommandState.getCurrentState(player.world.registryKey, server)
            return run(keepCommandState, context, player)
        }
        else {
            source.sendError(Text.literal("This command can only be used by a player"))
            return SINGLE_ERROR
        }
    }

    abstract fun run(
        keepCommandState: KeepCommandState,
        context: CommandContext<ServerCommandSource>,
        player: ServerPlayerEntity
    ): Int
}
